### PR TITLE
[PLAT-1640] Use custom audience validation implementation

### PR DESCRIPTION
**Pull request by sylank**
In the original implementation, the JWT library used `AND` based audience validation. In our implementation the library is using `OR` based audience validation

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+service/debug.test
```
Comment (zsolt-marta-bitrise) [inline .gitignore:1]: Let's exclude this from the PR :)

File: service/validator.go [hunk 2]
```diff
@@ -113,9 +129,63 @@ func (sv ValidatorConfig) ValidateRequestAndReturnToken(r *http.Request) (TokenW
...
+		return err
+	}
+
+	var audiencesInToken []string
+
+	switch v := payload["aud"].(type) {
+	default:
+		fmt.Printf("unexpected type %T", v)
+	case string:
+		audience, ok := payload["aud"].(string)
```
Comment (zsolt-marta-bitrise) [inline service/validator.go:153]: At this point you don't need the safe type assertion

File: service/validator.go [hunk 3]
```diff
@@ -113,9 +129,63 @@ func (sv ValidatorConfig) ValidateRequestAndReturnToken(r *http.Request) (TokenW
...
+		if !ok {
+			panic("type assertion failed")
+		}
+		audiencesInToken = []string{audience}
+	case []interface{}:
+		audienceInterface, ok := payload["aud"].([]interface{})
+		if !ok {
```
Comment (zsolt-marta-bitrise) [inline service/validator.go:160]: At this point you don't need the safe type assertion

File: service/validator.go [hunk 4]
```diff
@@ -113,9 +129,63 @@ func (sv ValidatorConfig) ValidateRequestAndReturnToken(r *http.Request) (TokenW
...
+		for i, v := range audienceInterface {
+			audience, ok := v.(string)
+			if !ok {
+				panic("type assertion failed")
+			}
+			audiencesInToken[i] = audience
+		}
+	}
+
+	if len(audiencesInToken) > 0 {
```
Comment (zsolt-marta-bitrise) [inline service/validator.go:173]: We need to validate the audiences in the config regardless whether the token has any. If we expect 2 audiences, but the token has none, it should fail

File: service/validator.go [hunk 5]
```diff
@@ -113,9 +129,63 @@ func (sv ValidatorConfig) ValidateRequestAndReturnToken(r *http.Request) (TokenW
...
+		found := false
+		for _, aud := range audiencesInToken {
```
Comment (zsolt-marta-bitrise) [inline service/validator.go:175]: Should be the other way around: for all expected tokens, if any of those are contained, it is successful.

**Review by zsolt-marta-bitrise — CHANGES_REQUESTED** [review]

File: service/validator.go (hunk 2 above)
Comment (imrekel) [reply service/validator.go:153]: are you sure? I mean **errcheck** fails if you don't check the result

File: service/validator.go (hunk 3 above)
Comment (imrekel) [reply service/validator.go:160]: are you sure? I mean **errcheck** fails if you don't check the result

File: service/validator.go (hunk 2 above)
Comment (zsolt-marta-bitrise) [reply service/validator.go:153]: You can use it like this if you assign the result to a var: https://stackoverflow.com/a/21934207/765062 
here, you can readily use the var, without further type assertion

File: service/validator.go (hunk 3 above)
Comment (zsolt-marta-bitrise) [reply service/validator.go:160]: See my other response

File: service/validator.go (hunk 2 above)
Comment (imrekel) [reply service/validator.go:153]: nice!

**Review by zsolt-marta-bitrise — APPROVED** [review]

**Pull request merged**